File: frontend/src/types/task.ts
export type TaskStatus = 'OPEN' | 'COMPLETED' | 'CANCELLED'

export type RelatedRecordType = 'LEAD' | 'ACCOUNT' | 'CONTACT' | 'DEAL'

/** Matches backend TaskResponse */
export interface Task {
  id: string
  relatedType: RelatedRecordType
  relatedId: string
  relatedName: string
  leadId?: string | null
  leadName?: string | null
  accountId?: string | null
  accountName?: string | null
  contactId?: string | null
  contactName?: string | null
  dealId?: string | null
  dealName?: string | null
  assignedToId: string
  assignedToName: string
  title: string
  description: string | null
  dueAt: string
  reminderAt: string | null
  status: TaskStatus
  createdAt: string
  updatedAt: string
}

/** Matches backend TaskCreateRequest */
export interface TaskCreateRequest {
  leadId?: string | null
  accountId?: string | null
  contactId?: string | null
  dealId?: string | null
  assignedToId?: string | null
  title: string
  description?: string | null
  dueAt: string
  reminderAt?: string | null
}

/** Matches backend TaskUpdateRequest */
export interface TaskUpdateRequest {
  leadId?: string | null
  accountId?: string | null
  contactId?: string | null
  dealId?: string | null
  assignedToId: string
  title: string
  description?: string | null
  dueAt: string
  reminderAt?: string | null
  status: TaskStatus
}

export const RELATED_RECORD_TYPES: RelatedRecordType[] = ['LEAD', 'ACCOUNT', 'CONTACT', 'DEAL']

export function relatedTypeLabel(type: RelatedRecordType): string {
  switch (type) {
    case 'LEAD':
      return 'Lead'
    case 'ACCOUNT':
      return 'Account'
    case 'CONTACT':
      return 'Contact'
    case 'DEAL':
      return 'Deal'
    default:
      return type
  }
}

export function formatRelatedRecord(task: Task): string {
  return `${relatedTypeLabel(task.relatedType)} · ${task.relatedName}`
}

export function relationFields(task: Task): Pick<TaskUpdateRequest, 'leadId' | 'accountId' | 'contactId' | 'dealId'> {
  return {
    leadId: task.leadId ?? null,
    accountId: task.accountId ?? null,
    contactId: task.contactId ?? null,
    dealId: task.dealId ?? null,
  }
}

File: backend/src/main/java/com/flowcrm/task/Task.java
package com.flowcrm.task;

import com.flowcrm.account.Account;
import com.flowcrm.contact.Contact;
import com.flowcrm.deal.Deal;
import com.flowcrm.enums.RelatedRecordType;
import com.flowcrm.enums.TaskStatus;
import com.flowcrm.lead.Lead;
import com.flowcrm.user.User;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "tasks")
public class Task {

    @Id
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "lead_id")
    private Lead lead;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "account_id")
    private Account account;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "contact_id")
    private Contact contact;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "deal_id")
    private Deal deal;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "assigned_to_id", nullable = false)
    private User assignedTo;

    @Column(nullable = false)
    private String title;

    @Column(length = 2000)
    private String description;

    @Column(name = "due_at", nullable = false)
    private Instant dueAt;

    @Column(name = "reminder_at")
    private Instant reminderAt;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 32)
    private TaskStatus status;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @PrePersist
    void onCreate() {
        Instant now = Instant.now();
        if (id == null) {
            id = UUID.randomUUID();
        }
        if (status == null) {
            status = TaskStatus.OPEN;
        }
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    void onUpdate() {
        updatedAt = Instant.now();
    }

    public void clearRelations() {
        this.lead = null;
        this.account = null;
        this.contact = null;
        this.deal = null;
    }

    public RelatedRecordType relatedType() {
        if (lead != null) {
            return RelatedRecordType.LEAD;
        }
        if (account != null) {
            return RelatedRecordType.ACCOUNT;
        }
        if (contact != null) {
            return RelatedRecordType.CONTACT;
        }
        if (deal != null) {
            return RelatedRecordType.DEAL;
        }
        throw new IllegalStateException("Task has no related record");
    }

    public UUID relatedId() {
        return switch (relatedType()) {
            case LEAD -> lead.getId();
            case ACCOUNT -> account.getId();
            case CONTACT -> contact.getId();
            case DEAL -> deal.getId();
        };
    }

    public String relatedName() {
        return switch (relatedType()) {
            case LEAD -> lead.getFullName();
            case ACCOUNT -> account.getName();
            case CONTACT -> (contact.getFirstName() + " " + contact.getLastName()).trim();
            case DEAL -> deal.getName();
        };
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public Lead getLead() {
        return lead;
    }

    public void setLead(Lead lead) {
        this.lead = lead;
    }

    public Account getAccount() {
        return account;
    }

    public void setAccount(Account account) {
        this.account = account;
    }

    public Contact getContact() {
        return contact;
    }

    public void setContact(Contact contact) {
        this.contact = contact;
    }

    public Deal getDeal() {
        return deal;
    }

    public void setDeal(Deal deal) {
        this.deal = deal;
    }

    public User getAssignedTo() {
        return assignedTo;
    }

    public void setAssignedTo(User assignedTo) {
        this.assignedTo = assignedTo;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Instant getDueAt() {
        return dueAt;
    }

    public void setDueAt(Instant dueAt) {
        this.dueAt = dueAt;
    }

    public Instant getReminderAt() {
        return reminderAt;
    }

    public void setReminderAt(Instant reminderAt) {
        this.reminderAt = reminderAt;
    }

    public TaskStatus getStatus() {
        return status;
    }

    public void setStatus(TaskStatus status) {
        this.status = status;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public Instant getUpdatedAt() {
        return updatedAt;
    }
}
